File: apps/api/src/ai_infra_api/schemas/api_resources.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, model_validator


class ProviderCapabilitiesResponse(BaseModel):
    credential_validation: bool
    model_discovery: bool
    balance_sync: bool
    usage_sync: bool
    usage_by_model: bool
    usage_by_credential: bool
    manual_usage_import: bool


class ApiProviderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    display_name: str
    provider_type: str
    default_base_url: str | None
    capabilities: ProviderCapabilitiesResponse
    is_enabled: bool


class ApiAccountCreate(BaseModel):
    provider_slug: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    purpose: str | None = Field(default=None, max_length=255)
    owner: str | None = Field(default=None, max_length=128)
    base_url: HttpUrl | None = None
    billing_currency: str | None = Field(default=None, max_length=16)
    monthly_budget: float | None = Field(default=None, ge=0)
    tags: list[str] = Field(default_factory=list, max_length=32)
    notes: str | None = Field(default=None, max_length=4_096)
    credential_name: str | None = Field(default=None, max_length=128)
    credential_type: str = "api_key"
    credential_value: str | None = Field(default=None, min_length=1, max_length=8_192)

    @model_validator(mode="after")
    def credential_fields_match(self) -> "ApiAccountCreate":
        if bool(self.credential_name) != bool(self.credential_value):
            raise ValueError("credential_name and credential_value must be provided together")
        return self


class ApiAccountUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    purpose: str | None = Field(default=None, max_length=255)
    owner: str | None = Field(default=None, max_length=128)
    base_url: HttpUrl | None = None
    status: str | None = None
    billing_currency: str | None = Field(default=None, max_length=16)
    monthly_budget: float | None = Field(default=None, ge=0)
    tags: list[str] | None = Field(default=None, max_length=32)
    notes: str | None = Field(default=None, max_length=4_096)


class ApiAccountResponse(BaseModel):
    id: uuid.UUID
    provider: ApiProviderResponse
    name: str
    purpose: str | None
    owner: str | None
    base_url: str
    status: str
    billing_currency: str | None
    monthly_budget: float | None
    tags: list[str]
    notes: str | None
    last_verified_at: datetime | None
    last_synced_at: datetime | None
    credential_count: int = 0
    model_count: int = 0
    latest_balance: float | None = None
    latest_usage_cost: float | None = None
    created_at: datetime
    updated_at: datetime


class ApiCredentialCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    credential_type: str = "api_key"
    value: str = Field(min_length=1, max_length=8_192)
    expires_at: datetime | None = None


class ApiCredentialUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    expires_at: datetime | None = None


class ApiCredentialRotate(BaseModel):
    value: str = Field(min_length=1, max_length=8_192)


class ApiCredentialResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID
    name: str
    credential_type: str
    masked_value: str
    status: str
    expires_at: datetime | None
    last_validated_at: datetime | None
    last_error_code: str | None
    last_error_message: str | None
    created_at: datetime
    updated_at: datetime


class ApiAccountModelManual(BaseModel):
    provider_model_id: str = Field(min_length=1, max_length=255)
    display_name: str | None = Field(default=None, max_length=255)
    model_family: str | None = Field(default=None, max_length=128)
    capabilities: list[str] = Field(default_factory=list, max_length=32)
    context_window: int | None = Field(default=None, ge=1)


class ApiAccountModelResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID
    provider_model_id: str
    display_name: str | None
    model_family: str | None
    capabilities: list[str]
    context_window: int | None
    is_available: bool
    source: str
    discovered_at: datetime
    last_seen_at: datetime


class ApiUsageManualCreate(BaseModel):
    period_start: datetime
    period_end: datetime
    granularity: str = "day"
    credential_id: uuid.UUID | None = None
    provider_model_id: str | None = Field(default=None, max_length=255)
    request_count: int | None = Field(default=None, ge=0)
    input_tokens: int | None = Field(default=None, ge=0)
    output_tokens: int | None = Field(default=None, ge=0)
    cached_tokens: int | None = Field(default=None, ge=0)
    total_tokens: int | None = Field(default=None, ge=0)
    cost_amount: float | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, max_length=16)
    metadata: dict[str, str | int | float | bool | None] = Field(default_factory=dict)

    @model_validator(mode="after")
    def period_is_valid(self) -> "ApiUsageManualCreate":
        if self.period_end <= self.period_start:
            raise ValueError("period_end must be after period_start")
        return self


class ApiUsageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID
    credential_id: uuid.UUID | None
    provider_model_id: str | None
    period_start: datetime
    period_end: datetime
    granularity: str
    request_count: int | None
    input_tokens: int | None
    output_tokens: int | None
    cached_tokens: int | None
    total_tokens: int | None
    cost_amount: float | None
    currency: str | None
    source: str
    collected_at: datetime


class ApiBalanceManualCreate(BaseModel):
    balance_amount: float | None = None
    credit_limit: float | None = None
    remaining_credit: float | None = None
    currency: str | None = Field(default=None, max_length=16)
    expires_at: datetime | None = None


class ApiBalanceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID
    balance_amount: float | None
    credit_limit: float | None
    remaining_credit: float | None
    currency: str | None
    expires_at: datetime | None
    source: str
    collected_at: datetime


class ApiSyncRequest(BaseModel):
    sync_types: list[str] = Field(min_length=1, max_length=3)


class ApiSyncRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID
    sync_type: str
    status: str
    requested_by_user_id: uuid.UUID | None
    started_at: datetime | None
    completed_at: datetime | None
    records_written: int
    error_code: str | None
    error_message: str | None
    details: dict[str, object]
    created_at: datetime
    updated_at: datetime


class ApiUsageSummaryResponse(BaseModel):
    account_count: int
    request_count: int
    input_tokens: int
    output_tokens: int
    total_tokens: int
    costs_by_currency: dict[str, float]

File: apps/web/src/lib/api/api-resources.ts
import { z } from "zod";

const nullableString = z.string().nullable();
const nullableNumber = z.number().nullable();

export const providerCapabilitiesSchema = z.object({
  credential_validation: z.boolean(),
  model_discovery: z.boolean(),
  balance_sync: z.boolean(),
  usage_sync: z.boolean(),
  usage_by_model: z.boolean(),
  usage_by_credential: z.boolean(),
  manual_usage_import: z.boolean(),
});

export const apiProviderSchema = z.object({
  id: z.string(),
  slug: z.string(),
  display_name: z.string(),
  provider_type: z.string(),
  default_base_url: nullableString,
  capabilities: providerCapabilitiesSchema,
  is_enabled: z.boolean(),
});

export const apiAccountSchema = z.object({
  id: z.string(),
  provider: apiProviderSchema,
  name: z.string(),
  purpose: nullableString,
  owner: nullableString,
  base_url: z.string(),
  status: z.string(),
  billing_currency: nullableString,
  monthly_budget: nullableNumber,
  tags: z.array(z.string()),
  notes: nullableString,
  last_verified_at: nullableString,
  last_synced_at: nullableString,
  credential_count: z.number().int(),
  model_count: z.number().int(),
  latest_balance: nullableNumber,
  latest_usage_cost: nullableNumber,
  created_at: z.string(),
  updated_at: z.string(),
});

export const apiCredentialSchema = z.object({
  id: z.string(),
  account_id: z.string(),
  name: z.string(),
  credential_type: z.string(),
  masked_value: z.string(),
  status: z.string(),
  expires_at: nullableString,
  last_validated_at: nullableString,
  last_error_code: nullableString,
  last_error_message: nullableString,
  created_at: z.string(),
  updated_at: z.string(),
});

export const apiAccountModelSchema = z.object({
  id: z.string(),
  account_id: z.string(),
  provider_model_id: z.string(),
  display_name: nullableString,
  model_family: nullableString,
  capabilities: z.array(z.string()),
  context_window: nullableNumber,
  is_available: z.boolean(),
  source: z.string(),
  discovered_at: z.string(),
  last_seen_at: z.string(),
});

export const apiUsageSchema = z.object({
  id: z.string(),
  account_id: z.string(),
  credential_id: nullableString,
  provider_model_id: nullableString,
  period_start: z.string(),
  period_end: z.string(),
  granularity: z.string(),
  request_count: nullableNumber,
  input_tokens: nullableNumber,
  output_tokens: nullableNumber,
  cached_tokens: nullableNumber,
  total_tokens: nullableNumber,
  cost_amount: nullableNumber,
  currency: nullableString,
  source: z.string(),
  collected_at: z.string(),
});

export const apiBalanceSchema = z.object({
  id: z.string(),
  account_id: z.string(),
  balance_amount: nullableNumber,
  credit_limit: nullableNumber,
  remaining_credit: nullableNumber,
  currency: nullableString,
  expires_at: nullableString,
  source: z.string(),
  collected_at: z.string(),
});

export const apiSyncRunSchema = z.object({
  id: z.string(),
  account_id: z.string(),
  sync_type: z.string(),
  status: z.string(),
  requested_by_user_id: nullableString,
  started_at: nullableString,
  completed_at: nullableString,
  records_written: z.number().int(),
  error_code: nullableString,
  error_message: nullableString,
  details: z.record(z.string(), z.unknown()),
  created_at: z.string(),
  updated_at: z.string(),
});

export const apiUsageSummarySchema = z.object({
  account_count: z.number().int(),
  request_count: z.number().int(),
  input_tokens: z.number().int(),
  output_tokens: z.number().int(),
  total_tokens: z.number().int(),
  costs_by_currency: z.record(z.string(), z.number()),
});

export type ApiProvider = z.infer<typeof apiProviderSchema>;
export type ApiAccount = z.infer<typeof apiAccountSchema>;
export type ApiCredential = z.infer<typeof apiCredentialSchema>;
export type ApiAccountModel = z.infer<typeof apiAccountModelSchema>;
export type ApiUsage = z.infer<typeof apiUsageSchema>;
export type ApiBalance = z.infer<typeof apiBalanceSchema>;
export type ApiSyncRun = z.infer<typeof apiSyncRunSchema>;
export type ApiUsageSummary = z.infer<typeof apiUsageSummarySchema>;

export const apiResourceQueryKeys = {
  all: ["api-resources"] as const,
  providers: ["api-resources", "providers"] as const,
  accounts: ["api-resources", "accounts"] as const,
  summary: ["api-resources", "summary"] as const,
  credentials: (accountId: string) => ["api-resources", accountId, "credentials"] as const,
  models: (accountId: string) => ["api-resources", accountId, "models"] as const,
  usage: (accountId: string) => ["api-resources", accountId, "usage"] as const,
  balance: (accountId: string) => ["api-resources", accountId, "balance"] as const,
  syncRuns: (accountId: string) => ["api-resources", accountId, "sync-runs"] as const,
};
